### ISSUE TITLE
DAG-MCNP assumes neutron as primary particle all the time- sets cell importances incorrectly

### ISSUE THREAD

**Issue by kkiesling** [svalinn/DAGMC#663]
**Describe the Bug**<br/>
If you run a DAG-MCNP 6.2 problem with weight windows using a primary mode particle that is not a neutron (eg setting `MODE p` for purely photons) the cell importances get set incorrectly. DAGMC, regardless of particle type, still sets the lcad file with neutron importances in the cells (`imp:n=1` or `imp:n=1`). This causes issues with MCNP then because it will then do the following sequence to set the importances of all the cells:
1. reads the `imp:n` designator from the lcad file and sets cell importances for the neutrons correctly (all volumes get importance 1 except the graveyard that gets importance 0).
2. Sees that we are in mode p for weight windows and tries to read the photon cell importances based on weight window information. Because no information is provided, it defaults all cell importances to 1 (including the graveyard).
3. Then goes back to the neutron importances to update them to match the photon importances because that is the weight window mode particle. After doing this all cells/volumes in the geometry are assigned importance of 1 for both photons and neutrons, including the graveyard volume.

These incorrect importances cause lost particles because they are not properly terminated when entering the graveyard.

[This](https://github.com/svalinn/DAGMC/blob/develop/src/mcnp/mcnp_funcs.cpp#L272) is where DAGMC default sets the importances of cells as neutron importances with no regard of mode specified in the MCNP input.

**To Reproduce**<br/>
Run a DAG-MCNP problem in pure photon mode (or just a non neutron mode) that includes weight windows.

**Expected Behavior**<br/>
What should happen is the cell importances written to the lcad file are `imp:p` if we are in a pure photon mode. Or rather they should match the particle mode specified by the MCNP input. This will cause MCNP to correctly read those importances in as the photon importances in step 1 listed above and not go to steps 2 or 3 to incorrectly overwrite the cell importances.

**Additional Context**<br/>
Discussed this with @gonuke and there are a few solutions to be considered:

- An immediate fix is to alter the lcad file, changing `imp:n` to `imp:p` in the cell lines and rerunning with the changed lcad. This is does work, but is not ideal for a long term solution.
- Read the MODE particle specified in the input file. This will be difficult because at the time that DAGMC has to read the geometry and create the lcad file, MCNP has not yet read in MODE card and so we don't yet have that information.
- Tag the geometry with a MODE metadata and read it in. If it's not supplied, assume mode n.
- Have as an MCNP input option for the user to specify the mode in the MCNP input file on the DAGMC card (would need to do both the MCNP MODE card and this option). Default again can be mode n.

**Comment by gonuke** [comment]
I don't think the last option helps either as we haven't ready the DAGMC card either.

Another option is to have the user tag the photon importance on all cells in the problem, but that seems more cumbersome than just tagging the mode on the whole geometry.

**Comment by makeclean** [comment]
Is this not actually an MCNP bug? Surely this would cause issues for a regular MCNP input deck? Create a small MCNP input deck with neutron importances but a photon run mode, do you get the same behaviour?

On 26 Feb 2020, at 21:51, Paul Wilson <notifications@github.com<mailto:notifications@github.com>> wrote:


I don't think the last option helps either as we haven't ready the DAGMC card either.

Another option is to have the user tag the photon importance on all cells in the problem, but that seems more cumbersome than just tagging the mode on the whole geometry.

**Comment by gonuke** [comment]
So.... arguably, MCNP should be more intelligent in transferring importances from neutrons to photons in a case like this, but I think they probably consider it user error.  One could argue that there is no "correct" behavior if the user has not provided importances for the primary particle in the problem.

That said, we trigger this ambiguity by applying uniform importances to all non-graveyard cells that are always and only defined on neutrons.  We'd be better off trying to mimic what a "good" user would do by matching it to the primary particle type of their mode.

**Comment by gonuke** [comment]
To another solution.... what if we change "graveyard" to be "graveyard:x" where "x" is the particle type?

Another contributor to this is our design choice to allow the user to specify no importances at all, except by declaring the graveyard, but we don't allow them do declare which particles die in the graveyard.  Maybe that's a middle ground without adding more tags?

Could even be "graveyard:n,p", i.e. listing multiple particles

**Comment by kkiesling** [comment]
> Another option is to have the user tag the photon importance on all cells in the problem, but that seems more cumbersome than just tagging the mode on the whole geometry.

I believe there is already the ability in DAGMC to tag volumes with importances? But that probably still assumes neutrons for the tagged importances.

**Comment by kkiesling** [comment]
> To another solution.... what if we change "graveyard" to be "graveyard:x" where "x" is the particle type?
> Could even be "graveyard:n,p", i.e. listing multiple particles

This would be my thought at an ideal solution. Call it whatever you want ("mode:n,p" or "graveyard"p" etc). But one tag on one volume (probably the graveyard because it is easy to identify) that lists the particle type in the problem. If no mode tag is present on the graveyard volume, default to neutrons. Basically, the information provided by a good user would be the same particle types listed in the "mode" card in the MCNP input file.

**Comment by kkiesling** [comment]
@makeclean I agree with @gonuke in that MCNP is probably buggy in how it handles the incorrect input supplied by a less than ideal user in this sense. But also the way we create the lcad file with DAGMC currently mimics this less than ideal behavior. And if we know better because we are all good MCNP users (the best users, perfect MCNP users :laughing:), we should correct the lcad file so it provides the correct input to MCNP.

**Comment by gonuke** [comment]
> > Another option is to have the user tag the photon importance on all cells in the problem, but that seems more cumbersome than just tagging the mode on the whole geometry.
> 
> I believe there is already the ability in DAGMC to tag volumes with importances? But that probably still assumes neutrons for the tagged importances.

No - here, we definitely get the particle type from the tag name
